### PR TITLE
langchain[patch]: Add possibility to rerank retrieved docs in ParentDocumentRetriever and MultiQueryRetriever

### PR DESCRIPTION
**Pull request by karol-f**
With many documents from the vector store that are passed to LLM, final answers sometimes consist of information from
irrelevant documents, making it less precise and sometimes incorrect. Also, passing multiple irrelevant documents makes it
more expensive.
So there are two reasons to use rerank - precision and costs.

_Side note: there is currently Cohere SDK error on some environments with new Node/Next - https://github.com/cohere-ai/cohere-typescript/issues/136. It should be solved in the upcoming release of Cohere SDK._

### CODE REVIEW

File: langchain/src/retrievers/parent_document.ts [hunk 1]
```diff
@@ -81,6 +84,10 @@ export class ParentDocumentRetriever extends MultiVectorRetriever {
     | VectorStoreRetrieverInterface<VectorStoreInterface>
     | undefined;
 
+  documentCompressor: BaseDocumentCompressor | undefined;
+
+  protected documentCompressorrMinRelevanceScore?: number = 0.4;
```
Comment (jacoblee93) [inline langchain/src/retrievers/parent_document.ts:89]: One `r` here

Comment (karol-f) [reply langchain/src/retrievers/parent_document.ts]: You are right, fixed!

File: examples/src/retrievers/parent_document_retriever_rerank.ts [hunk 2]
```diff
@@ -0,0 +1,66 @@
+import { OpenAIEmbeddings } from "@langchain/openai";
+import { CohereRerank } from "@langchain/cohere";
```
Comment (jacoblee93) [inline examples/src/retrievers/parent_document_retriever_rerank.ts:2]: So this doesn't yet implement `BaseDocumentCompressor` - I will fix that before merging this.

Comment (karol-f) [reply examples/src/retrievers/parent_document_retriever_rerank.ts:2]: Thank you

**Comment by jacoblee93** [comment]
Really sorry for the delay - I don't think I'll have time to implement this for Cohere soon.

If you'd like, we can remove the docs/example and get this merged sooner?

**Comment by karol-f** [comment]
@jacoblee93 Sounds perfect. I've removed examples. Regards!

**Comment by jacoblee93** [comment]
Hi really sorry for the delay - @bracesproul or I will address this on Monday now that DocumentCompressor is in core

**Comment by karol-f** [comment]
> Hi really sorry for the delay - @bracesproul or I will address this on Monday now that DocumentCompressor is in core

Fantastic!

File: examples/src/retrievers/parent_document_retriever_rerank.ts (hunk 2 above)
Comment (karol-f) [reply examples/src/retrievers/parent_document_retriever_rerank.ts:2]: Hi, I've changed a code so you can feed any document compressor to both Parent Retreiver and Multi Query Retreiver so it's much cleaner and it's up to developer what to use.

**Comment by karol-f** [comment]
@jacoblee93 I've refactored the above code, and now it's document compressor agnostic so you can use Cohere reranking (as shown in the example) or any other solution that fit's you.

Tested on the real app using NPM library Patch Package.

**Comment by jacoblee93** [comment]
Sorry about the delay!

**Comment by jacoblee93** [comment]
Thank you for your patience!

**Comment by karol-f** [comment]
Thank you for the good work. You are awesome maintainer!

**Pull request merged**